synapse/storage/databases/main/event_push_actions.py

```diff
@@ -1106,13 +1222,18 @@ def _handle_new_receipts_for_notifs_txn(self, txn: LoggingTransaction) -> bool:
         )
         rows = txn.fetchall()
 
-        # For each new read receipt we delete push actions from before it and
-        # recalculate the summary.
-        for _, room_id, user_id, stream_ordering in rows:
+        # First handle all the rows without a thread ID (i.e. ones that apply to
+        # the entire room).
+        for _, room_id, user_id, thread_id, stream_ordering in rows:
             # Only handle our own read receipts.
             if not self.hs.is_mine_id(user_id):
                 continue
 
+            if thread_id is not None:
+                continue
```

Comment:
The comment above the `for` loop already explains the following but similar to the `is_mine_id` skip condition above, it's nice not to have to reason the negative condition yourself.

```suggestion
            # Skip any read receipts that belong to a thread in this first loop.
            # We are only concerned with unthreaded read receipts here.
            if thread_id is not None:
                continue
```

---

As mentioned in https://github.com/matrix-org/synapse/pull/13877#discussion_r977923893, splitting the list into `unthreaded_receipts` and `threaded_receipts` would also make this clear and solve this problem 👍